### ISSUE TITLE
SC2236 with unquoted conditions

### ISSUE THREAD

**Issue by jkroepke** [koalaman/shellcheck#1327]
#### For bugs
- Rule Id (if any, e.g. SC1000):  SC2236
- My shellcheck version (`shellcheck --version` or "online"): onine
- [x] I read the issue's wiki page, e.g. https://github.com/koalaman/shellcheck/wiki/SC2086
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

if [ ! -z ${var+x} ]; then
  echo .
fi
```

#### Here's what shellcheck currently says:

```sh
Line 3:
if [ ! -z ${var+x} ]; then
     ^-- SC2236: Use -n instead of ! -z.
          ^-- SC2154: var is referenced but not assigned.
```

#### Here's what I wanted or expected to see:
```sh
Line 3:
if [ ! -z ${var+x} ]; then
          ^-- SC2154: var is referenced but not assigned.
```

Because if i'm using -n, I got an another issue:

```sh
#!/bin/bash

if [ -n ${var+x} ]; then
  echo .
fi
```
```
Line 3:
if [ -n ${var+x} ]; then
        ^-- SC2070: -n doesn't work with unquoted arguments. Quote or use [[ ]].
        ^-- SC2154: var is referenced but not assigned.
```

**Comment by matthewpersico** [comment]
If -n doesn't work with unquoted args, wouldn't -z also not work? The ! -z to -n swap was a recent addition; I contend that we should be seeing:

    if [ ! -z ${var+x} ]; then
         ^-- SC2236: Use -n instead of ! -z.
            ^-- SC2070: -n and -z don't work with unquoted arguments. Quote or use [[ ]].
                ^-- SC2154: var is referenced but not assigned.

**Comment by jkroepke** [comment]
Maybe -z/-n requires quotes unless `var` contains one char. in this example `var` is forced by the expression `+x` into one char. 

SC2070 should not appear if there this a expression like `${var+z}`

https://stackoverflow.com/a/13864829

**Comment by Tamaranch** [comment]
Incidentally, since `${var+x}` is empty if and only if `${var}` is empty, why ShellCheck doesn't suggest to replace all of this by `if [ -n "${var}" ]; …`, or to use `${var+x}` elsewhere without a redundant test?

**Comment by jkroepke** [comment]
If you have `set -u` / `set -o nounset` enabled, `if [ -n "${var}" ]; …` throws an error while `if [ -n "${var+x}" ]; …` not.

There are pros and cons about this way.

**Comment by Tamaranch** [comment]
Yes, it's true. Despite this, there remains some redundancy, but perhaps not always reducible…

**Comment by koalaman** [comment]
ShellCheck is right. You should be using `-n` and you should be quoting it.

`[ ! -z ${var+x} ]` works mostly by accident because when the variable is unset, it becomes `[ ! -z ]` which is equivalent to `[ ! -n "-z" ]` which is false, and that happens to be the expected result. It's clever but it's not good practice.
